cobra/medium/minimal_medium.py

```diff
@@ -0,0 +1,229 @@
+# -*- coding: utf-8 -*-
+
+"""Contains functions and helpers to obtain minimal growth media."""
+
+from optlang.symbolics import Zero
+from optlang.interface import OPTIMAL
+import numpy as np
+import pandas as pd
+import logging
+from cobra.medium.boundary_types import find_boundary_types
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+def add_linear_obj(model):
+    """Add a linear version of a minimal medium to the model solver.
+
+    Changes the optimization objective to finding the growth medium requiring
+    the smallest total import flux::
+
+        minimize sum |r_i| for r_i in import_reactions
+
+    Arguments
+    ---------
+    model : cobra.Model
+        The model to modify.
+    """
+    coefs = {}
+    for rxn in find_boundary_types(model, "exchange"):
+        export = len(rxn.reactants) == 1
+        if export:
+            coefs[rxn.reverse_variable] = 1
+        else:
+            coefs[rxn.forward_variable] = 1
+    model.objective.set_linear_coefficients(coefs)
+    model.objective.direction = "min"
+
+
+def add_mip_obj(model):
+    """Add a mixed-integer version of a minimal medium to the model.
+
+    Changes the optimization objective to finding the medium with the least
+    components::
+
+        minimize size(R) where R part of import_reactions
+
+    Arguments
+    ---------
+    model : cobra.model
+        The model to modify.
+    """
+    if len(model.variables) > 1e4:
+        LOGGER.warning("the MIP version of minimal media is extremely slow for"
+                       " models that large :(")
+    exchange_rxns = find_boundary_types(model, "exchange")
+    M = max(np.max(np.abs(r.bounds)) for r in exchange_rxns)
```

Comment:
I think numpy is overkill here. (It will convert each tuple into a new array.) In the following `bounds` is an array of pairs.

```python
In [10]: %timeit max(np.max(np.abs(p)) for p in bounds)
9.2 ms ± 67 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

```python
In [13]: %timeit max(abs(x) for p in bounds for x in p)
421 µs ± 2.66 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```